elf_reader.go

```diff
@@ -479,76 +475,118 @@ func mapSpecFromBTF(spec *btf.Spec, name string) (*MapSpec, error) {
 	if err != nil {
 		return nil, fmt.Errorf("can't get size of BTF value: %w", err)
 	}
-	valueSize := uint32(size)
 
-	var (
-		mapType, flags, maxEntries uint32
-		pinType                    PinType
-	)
+	mapSpec := &MapSpec{
+		Name:      SanitizeName(name, -1),
+		ValueSize: uint32(size),
+		KeySize:   keySize,
+	}
+
 	for _, member := range btfMapMembers {
 		switch member.Name {
 		case "type":
-			mapType, err = uintFromBTF(member.Type)
+			mapType, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get type: %w", err)
 			}
+			mapSpec.Type = MapType(mapType)
 
 		case "map_flags":
-			flags, err = uintFromBTF(member.Type)
+			flags, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get BTF map flags: %w", err)
 			}
+			mapSpec.Flags = flags
 
 		case "max_entries":
-			maxEntries, err = uintFromBTF(member.Type)
+			maxEntries, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get BTF map max entries: %w", err)
 			}
+			mapSpec.MaxEntries = maxEntries
 
 		case "key_size":
 			if _, isVoid := keyType.(*btf.Void); !isVoid {
 				return nil, errors.New("both key and key_size given")
 			}
 
-			keySize, err = uintFromBTF(member.Type)
+			keySize, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get BTF key size: %w", err)
 			}
+			mapSpec.KeySize = keySize
 
 		case "value_size":
 			if _, isVoid := valueType.(*btf.Void); !isVoid {
 				return nil, errors.New("both value and value_size given")
 			}
 
-			valueSize, err = uintFromBTF(member.Type)
+			valueSize, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get BTF value size: %w", err)
 			}
+			mapSpec.ValueSize = valueSize
 
 		case "pinning":
+			if isInnermap {
+				return nil, fmt.Errorf("inner def can't be pinned")
+			}
+
 			pinning, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get pinning: %w", err)
 			}
 
-			pinType = PinType(pinning)
+			mapSpec.Pinning = PinType(pinning)
+
+		case "values":
+			if isInnermap {
+				return nil, fmt.Errorf("multi-level inner maps not supported")
+			}
+
+			innerMapStruct, err := arrayFromBTF(member.Type)
+			if err != nil {
+				return nil, fmt.Errorf("can't get BTF map values: %w", err)
+			}
+
+			if mapSpec.Type != ArrayOfMaps && mapSpec.Type != HashOfMaps {
+				return nil, fmt.Errorf("should be map-in-map")
+			}
+
+			inner, err := mapSpecFromBTF(spec, string(innerMapStruct.Name), maps, true)
```

Comment:
I think this call doesn't work: the inner map can be defined using an anonymous struct, so there won't be a name you can use to look it up.

Please double check, but I think `innerMapStruct` already contains the map definition. You probably need to split `btf.Spec.Map` into two functions: one that uses `name` to find `mapStruct` and one that turns `mapStruct` into `btf.Map`. I think this function `mapSpecFromBTF` might also have to be split to make things work.

Reply:
Let's check again in detail.
But now a simple print appears to be working because the name is displayed.
```
	if isInnermap {
		fmt.Println("check inner ******")
		fmt.Println(name)
	}
```
Can you elaborate on what kind of case it is not working?

Reply:
For example (from #155):

```
struct {
	__uint(type, BPF_MAP_TYPE_ARRAY_OF_MAPS);
	__uint(max_entries, MAX_ENTRIES);
	__type(key, __u32);
	__type(value, __u32);
	__array(values, struct {
		__uint(type, BPF_MAP_TYPE_ARRAY);
		__uint(max_entries, 1);
		__type(key, __u32);
		__type(value, __u32);
	});
}
```

Here, `values` has a type that is an anonymous struct. `innerMapStruct.Name` is therefore `""`.

Reply:
!!!!!
I know! i will fix it 🙆‍♀️